harness/determined/common/experimental/trial.py

```diff
@@ -165,59 +196,84 @@ def select_checkpoint(
+        """
+
+        if (sort_by is None) != (order_by is None):
+            raise AssertionError("sort_by and order_by must be set together")
+
+        def get_trial_checkpoints(offset: int) -> bindings.v1GetTrialCheckpointsResponse:
             return bindings.get_GetTrialCheckpoints(
                 self._session,
                 id=self.id,
-                orderBy=bindings.v1OrderBy.ORDER_BY_DESC,
-                sortBy=bindings.v1GetTrialCheckpointsRequestSortBy.SORT_BY_BATCH_NUMBER,
+                orderBy=order_by and order_by._to_bindings(),
+                sortBy=sort_by._to_bindings() if isinstance(sort_by, CheckpointSortBy) else None,
                 offset=offset,
             )
 
-        resps = api.read_paginated(get_one)
+        resps = api.read_paginated(get_trial_checkpoints)
 
         checkpoints = [
             checkpoint.Checkpoint._from_bindings(c, self._session)
             for r in resps
             for c in r.checkpoints
         ]
 
-        if not checkpoints:
-            raise AssertionError("No checkpoint found for trial {}".format(self.id))
-
-        if latest:
-            return checkpoints[0]
+        # If sort_by was a defined field, we already sorted and ordered.
+        if isinstance(sort_by, CheckpointSortBy) or not checkpoints:
+            return checkpoints
 
+        # If sort not specified, sort and order default to searcher configs.
         if not sort_by:
             training = checkpoints[0].training
             assert training
             config = training.experiment_config
-            sb = config.get("searcher", {}).get("metric")
-            if not isinstance(sb, str):
+            searcher_metric = config.get("searcher", {}).get("metric")
+            if not isinstance(searcher_metric, str):
                 raise ValueError(
                     "no searcher.metric found in experiment config; please provide a sort_by metric"
                 )
-            sort_by = sb
+            sort_by = searcher_metric
             smaller_is_better = config.get("searcher", {}).get("smaller_is_better", True)
+            order_by = CheckpointOrderBy.ASC if smaller_is_better else CheckpointOrderBy.DESC
 
-        def has_metric(c: checkpoint.Checkpoint) -> bool:
-            if c.training is None:
-                return False
-            return sort_by in c.training.validation_metrics.get("avgMetrics", {})
-
-        checkpoints_with_metric = [c for c in checkpoints if has_metric(c)]
-
-        if not checkpoints_with_metric:
-            raise AssertionError(f"No checkpoint for trial {self.id} has metric {sort_by}")
-
-        best_checkpoint_func = min if smaller_is_better else max
+        assert sort_by is not None and order_by is not None, "sort_by and order_by not defined."
 
         def key(ckpt: checkpoint.Checkpoint) -> Any:
             training = ckpt.training
             assert training
             return training.validation_metrics["avgMetrics"][sort_by]
```

Comment:
Might also want to consider unusual cases (possible at least in Core API) where different checkpoints may be reported with validation metrics that have different keys across checkpoints. 

I think in these cases it might just be best to tell the user to sort chronologically, though.